### PR TITLE
Neighbors and Wine lesson corrected

### PR DESCRIPTION
**Pull request by YaroslavMayorov**

### CODE REVIEW

File: Neighbors and Wine/K-nearest neighbors/metric_classification.py [hunk 1]
```diff
@@ -2,9 +2,15 @@
 
 
 def knn(X_train, y_train, X_test, k, dist):
+    # The function will return the class for x based on its neighbors from the X_train
```
Comment (stephen-hero) [inline Neighbors and Wine/K-nearest neighbors/metric_classification.py]: The function predicts the class for x based on its neighbors within the X_train

File: Neighbors and Wine/K-nearest neighbors/metric_classification.py [hunk 2]
```diff
@@ -2,9 +2,15 @@
...
+    # sample.
     def classify_single(x):
+        # Here we create an array of distances from x to each of the X_train objects.
         dists = [dist(x, i) for i in X_train]
+        # This array will contain the indices of k nearest to the x objects.
         indices = np.argpartition(dists, k)[:k]
+        # The function returns the most frequent class among those in y_train represented
```
Comment (stephen-hero) [inline Neighbors and Wine/K-nearest neighbors/metric_classification.py]: The function returns the most frequent class among the labels in y_train identified

File: Neighbors and Wine/Leave-one-out/crossvalidation.py [hunk 3]
```diff
@@ -4,13 +4,24 @@
 
 def loocv(X_train, y_train, dist):
     def loo(k):
+        # This is a counter keeping track of the wrongly determined classes.
```
Comment (stephen-hero) [inline Neighbors and Wine/Leave-one-out/crossvalidation.py]: This is a counter for incorrectly predicted classes.

File: Neighbors and Wine/Leave-one-out/crossvalidation.py [hunk 4]
```diff
@@ -4,13 +4,24 @@
...
         c = 0
         for i in range(len(X_train)):
+            # Here we should choose the current training sample of objects
+            # including all but the one with index i.
```
Comment (stephen-hero) [inline Neighbors and Wine/Leave-one-out/crossvalidation.py]: by excluding the object at index i.

File: Neighbors and Wine/Leave-one-out/crossvalidation.py [hunk 5]
```diff
@@ -4,13 +4,24 @@
...
             x_train_cur = np.vstack([X_train[:i], X_train[i + 1:]])
+            # Here we should choose the current training sample of classes
+            # including all but the one with index i.
             y_train_cur = np.concatenate((y_train[:i], y_train[i + 1:]))
+            # Here we should place a condition when the algorithm trained on the samples
```
Comment (stephen-hero) [inline Neighbors and Wine/Leave-one-out/crossvalidation.py]: Here, we add a condition to check if the algorithm trained on the samples

File: Neighbors and Wine/Leave-one-out/task.md [hunk 6]
```diff
@@ -50,5 +50,22 @@ The function needs to take the training sample and the distance function:
 
 The function template is in the `crossvalidation.py` file. 
 
-Assess the precision and recall of the classifier with the optimal `k` and any two distance functions. 
-To do that, run `task.py`.
+Assess the precision and recall of the classifier with the optimal `k` and any two distance functions.
```
Comment (stephen-hero) [inline Neighbors and Wine/Leave-one-out/task.md]: Assess the precision and recall of the classifier using the optimal `k` and two different distance functions.

File: Neighbors and Wine/Leave-one-out/task.py [hunk 7]
```diff
@@ -8,7 +8,19 @@
...
     indices = np.random.permutation(X.shape[0])
     train_len = int(X.shape[0] * ratio)
-    return X[indices[:train_len]], y[indices[:train_len]], X[indices[train_len:]], y[indices[train_len:]]
+    # A slice with first train_len elements of X will be
+    # the training sample for objects.
+    X_train = X[indices[:train_len]]
+    # A slice with the rest of X will be
+    # the testing sample for objects.
+    X_test = X[indices[train_len:]]
+    # A slice with first train_len of y will be
```
Comment (stephen-hero) [inline Neighbors and Wine/Leave-one-out/task.py]: The first train_len elements of y will be

File: Neighbors and Wine/Precision and recall/task.md [hunk 8]
```diff
@@ -33,6 +33,7 @@ The function may look as follows:
             
         return class_precision_recall
 
+Don't forget to import the implemented function to `task.py` in order to use it in the main program and output the result.
```
Comment (stephen-hero) [inline Neighbors and Wine/Precision and recall/task.md]: Ensure you import the implemented function into `task.py` to use it in the main program and output the result.

File: Neighbors and Wine/Precision and recall/task.md [hunk 9]
```diff
@@ -51,3 +52,8 @@ or
 </div>
 
 Here we use the <a href="https://numpy.org/doc/1.18/reference/generated/numpy.unique.html">numpy.unique</a> function, which returns all unique elements of an array. It helps us get the list of all classes represented in the sample.
+
+To see the results of your code's work, you can add the following line to the `main` block in `task.py` block and run it:
```
Comment (stephen-hero) [inline Neighbors and Wine/Precision and recall/task.md]: To see your code in action, add the following line to the `main` block in `task.py` and run it:

File: Neighbors and Wine/Precision and recall/task.py [hunk 10]
```diff
@@ -7,7 +7,19 @@
 def train_test_split(X, y, ratio=0.8):
     indices = np.random.permutation(X.shape[0])
     train_len = int(X.shape[0] * ratio)
-    return X[indices[:train_len]], y[indices[:train_len]], X[indices[train_len:]], y[indices[train_len:]]
+    # A slice with first train_len elements of X will be
+    # the training sample for objects.
+    X_train = X[indices[:train_len]]
+    # A slice with the rest of X will be
```
Comment (stephen-hero) [inline Neighbors and Wine/Precision and recall/task.py]: The remaining elements of X will be

**Review by stephen-hero — APPROVED** [review]
